development-styleguide.md

```diff
@@ -0,0 +1,202 @@
+
+#### Avoid object modification.
+
+- Due to its dynamic nature, JavaScript makes it extremely easy to modify objects that you do not own
+- Such practice should generally be avoided as it can lead to maintenance problems and can produce hard to find bugs
+- If you don't own it, don't modify it
+
+
+```typescript
+let bob = { name : "Bob", surname: "Smith", age : 23 }
+bob.age = 24; // avoid
+// create a new object using spread operator to update only specific fields
+const olderBob = { ...bob, age: 24 }
+```
+
+### [FP-TS](https://gcanti.github.io/fp-ts/modules/Option.ts.html)
+
+#### Use [Options](https://github.com/gcanti/fp-ts/blob/1.12.0/src/Option.ts) to handle nullable values.
+
+```typescript
+import { fromNullable, Option, some } from "fp-ts/lib/Option";
+
+// this is a common scenario where we want to model an user with email preferences
+
+// undefined -> user doesn't give any feedback
+// true -> users wants to receive email
+// false -> users doesn't want to receive email
+type EmailPreference = boolean | undefined;
+
+type User = {
+  id: number;
+  emailPreference: EmailPreference;
+};
+
+const user: User = { id: 1, emailPreference: true };
+
+// check if user wants to receive email
+if (user.emailPreference !== undefined && user.emailPreference === true) {
+  // send email
+} else {
+  console.log("can't send email due to the user's preference");
+}
+
+// use option
+fromNullable(user.emailPreference).map(pref => {
+  if (pref) {
+    // send email
+  } else {
+    console.log("can't send email due to the user's preference");
+  }
+});
+```
+
+#### Use [Either](https://github.com/gcanti/fp-ts/blob/1.12.0/src/Either.ts) to handle success or failure.
+
+For some scenario, you may be interested why a computation failed.
+Using _Option_ is useful when you want to know if a value is present or not.
+The reason why it is not present it's not interesting.
+If you want to know more about computation failure you should use Either.
+
+Consider this example
+
+> write a function to validate a password
+>
+> it has to be:
+>
+> length >= 8
+>
+> contain at least a char in upper case
+>
+> contain at least a number
+
+#### a possible implementaion
+```typescript
+// return the error message if the password is not compliant, null otherwise
+export const validatePassword = (pwd: string): string | null => {
+  if (pwd.length < 8) {
+    return "at least 8 characters";
+  }
+  if (!/[A-Z]/g.test(pwd)) {
+    return "at least one capital letter";
+  }
+  if (!/[0-9]/g.test(pwd)) {
+    return "at least one number";
+  }
+  return null;
+};
+
+const myPws = "123456A";
+
+const validatePwd = validatePassword(myPws);
+if (myPws === null) {
+  console.log("password is ok");
+} else {
+  console.error(`the password is too weak: ${validatePwd}`);
+}
+```
+#### using Either
+```typescript
+// return the error if the password is not compliant, the password otherwise
+export const validatePassword = (pwd: string): Either<Error, string> => {
```

Comment:
this example is better handled with `Validation`  (accumulating errors),
see https://dev.to/gcanti/getting-started-with-fp-ts-either-vs-validation-5eja

Reply:
Since **[getValidation](https://github.com/gcanti/fp-ts/blob/master/src/Either.ts)** is available from version 2.0.0/typescript 3.5.+ (almost of our projects are not using the latest version) and I found that example too complex for this scope (validation and semigroup) I arranged the code to a basic example but I left the link for these cases where developers wants handle multiple errors

Do you think could it be enough?

Reply:
right !